### PR TITLE
Production-ready React module setup

### LINKED ISSUE #11: Publish to npm?
Hi Cheng,

Have you considered publishing to npm (even if its internals are still to change?). The npm package `react-animation` [is already taken](https://www.npmjs.com/package/react-animation). However, that other project [seems like it never kicked off](https://github.com/akiran/react-animation). Perhaps you can claim it or publish to a different name?

Thanks :)
Darío

### PR DESCRIPTION
**Pull request by nkbt**
1. I've added ESLint with pretty strict config. I updated config to match your code style. Fixed only some actual errors and minor style inconsistency
2. Added karma with one empty test. It also gives you a coverage
3. Split webpack config to production and development for easier support and extension in a future
4. Added ESLint loader, so it will throw warnings/errors while you develop things in dev mode
5. changed __DEV__ to process.env.NODE_ENV, since the same const is used in React itself to check if PropTypes checks should be done
6. Added bunch of scripts including prepublish, so you can publish module on npm now already built properly and without any extras (npmignore)

Should pretty much fix #11

### CODE REVIEW

**Comment by nkbt** [comment]
I tested all demos to make sure nothing is broken.

Next: I want to make demos built in a way you can easily do development and also publish gh-pages with no pain. Should be pretty quick (did not do it to keep this PR bit smaller)

File: .npmignore [hunk 1]
```diff
@@ -1,2 +1,13 @@
 react-animation.sublime-*
 server.js
+coverage/
+example/
+src/
+test/
+.coveralls.yml
```
Comment (nkbt) [inline .npmignore]: I recommend to add Coveralls to show test coverage. It is quite easy to setup

File: karma.conf.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+var webpackConfig = {
+  devtool: 'eval',
+  resolve: {
+    extensions: ['', '.js'],
+  },
+  module: {
+    loaders: process.env.COVERAGE ?
+      [
+        {test: /\.js$/, loader: 'babel', include: [path.resolve('./test')]},
+        {test: /\.js$/, loader: 'isparta', include: [path.resolve('./src')]},
```
Comment (nkbt) [inline karma.conf.js]: This is necessary to get proper ES6 code coverage

File: package.json [hunk 3]
```diff
@@ -2,22 +2,64 @@
...
+  "homepage": "https://github.com/chenglou/react-animation",
   "dependencies": {
     "react": ">=0.13.2"
   },
   "devDependencies": {
+    "babel": "^5.6.14",
+    "babel-core": "^5.6.15",
+    "babel-loader": "^5.3.1",
+    "coveralls": "^2.11.2",
     "css-layout": "0.0.2",
```
Comment (nkbt) [inline package.json]: Did not remove this one, but I reckon it is not used.

File: server.js [hunk 4]
```diff
@@ -4,8 +4,12 @@ var config = require('./webpack.config');
 
 new WebpackDevServer(webpack(config), {
   publicPath: config.output.publicPath,
-  hot: true
-}).listen(3000, 'localhost', function (err, result) {
+  hot: true,
+  stats: {
+    chunkModules: false,
```
Comment (nkbt) [inline server.js]: do not spit hundreds of modules into console on build, only essential data

File: src/Spring.js [hunk 5]
```diff
@@ -153,7 +157,7 @@ function mergeDiff(collA, collB, onRemove, accum) {
 }
 
 function mergeDiffObj(a, b, onRemove) {
-  let keys = mergeDiff(Object.keys(a), Object.keys(b), a => !onRemove(a), []);
+  let keys = mergeDiff(Object.keys(a), Object.keys(b), _a => !onRemove(_a), []);
```
Comment (nkbt) [inline src/Spring.js]: overrides function argument. here and after - I just added `_`, but ideally they should have some other name that will not clash.

File: src/Spring.js [hunk 6]
```diff
@@ -215,7 +220,7 @@ export default React.createClass({
   },
 
   componentDidMount() {
-    if (__DEV__) {
+    if (process.env.NODE_ENV !== 'production') {
```
Comment (nkbt) [inline src/Spring.js]: Same as in react itself

File: src/Spring.js [hunk 7]
```diff
@@ -284,38 +289,38 @@ export default React.createClass({
       this.state.currV = currV;
     }
 
-    if(hackOn != null) {
+    if (hackOn != null) {
       let {endValue} = this.props;
-      return <div {...this.props}>{
+      return (<div {...this.props}>{
```
Comment (nkbt) [inline src/Spring.js]: multi-line component not wrapped in `()`

File: src/Spring.js [hunk 8]
```diff
@@ -454,7 +460,8 @@ export let TransitionSpring = React.createClass({
           currV[key] = mapTree(zero, currVals[key]);
         });
 
-      let frameRate = now ? (Date.now() - now) / 1000 : FRAME_RATE;
+      // TODO: remove, unused.
+      // let frameRate = now ? (Date.now() - now) / 1000 : FRAME_RATE;
```
Comment (nkbt) [inline src/Spring.js]: this one is unused, and when removed - `now` var becomes unused too. So removed.

File: src/index.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import Spring, {TransitionSpring} from './Spring';
+import * as utils from './utils';
+import stepper from './stepper';
+
+
+export default {
```
Comment (nkbt) [inline src/index.js]: Export all things. Feel free to limit what is exported.

File: test/Spring-test.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
+describe('Spring', () => {
+  const SpringInjector = require('inject!../src/Spring');
+  let stepper, utils, Spring;
+
+
+  beforeEach(() => {
+    stepper = jasmine.createSpy('stepper');
```
Comment (nkbt) [inline test/Spring-test.js]: Mock all internal dependencies.

Not sure about these tools, maybe we don't need to mock them. I personally prefer to mock all internal libs and utils though.

File: webpack.config.js [hunk 11]
```diff
@@ -1,62 +1,84 @@
...
-    loaders: [{
-      test: /\.jsx?$/,
-      exclude: /build|node_modules/,
-      loaders: loaders
-    }]
+    loaders: [
+      {test: /\.js$/, loaders: ['react-hot', 'babel'], include: sources},
+    ],
+    preLoaders: [
+      {test: /\.js$/, loaders: ['eslint'], include: sources},
```
Comment (nkbt) [inline webpack.config.js]: process all files with eslint on compile step

**Comment by chenglou** [comment]
Hooooly s*** that's hardcore =D

Thanks a ton. I have some local changes (since two days ago... sorry for the delay) to the API that I want to push first. This might not merge cleanly afterward. Anyhow, I think this PR should be split. More on that after I'm done (soon).

**Comment by nkbt** [comment]
No worries, let me know how you want to split it. I will fix conflicts for
sure.

**Comment by nkbt** [comment]
Let's make it step by step. ESLint is completely independent #17. Configured for your code style. There are lots of syntax errors and couple of missing functions though. Did not fix them in that PR

**Comment by nkbt** [comment]
Ouch, just realized I accidentally made a PR from my master, not from the branch. After rebase PR has gone =(.

**Comment by chenglou** [comment]
Man, thanks again for all your effort and sorry for changing the files so quickly. I'll comment on the other issues...

**Comment by nkbt** [comment]
No worries. This is something I am really happy to do

**Pull request closed without merging**